### ISSUE TITLE
Carbon::parse('2020-04-110') => Warning instead of throwing Exception

### ISSUE THREAD

**Issue by opensourcerules** [briannesbitt/Carbon#2075]
Hello,

I encountered an issue with the following code:
```php
echo Carbon::parse('2020-04-110')->format('Y-m-d');
```

Carbon version: 2.33

PHP version: 7.22 // Edit 7.2.20

I expected to get:

```
Exception from vendor/nesbot/carbon/src/Carbon/Traits/Creator.php:179
```

But I actually get:

```
{"exception":"Error","message":"Call to a member function format() on boolean","code":0}
[Tue May  5 17:20:21 2020] PHP  17. Carbon\Carbon::parse() ...
[Tue May  5 17:20:21 2020] PHP  18. Carbon\Carbon::rawParse() \vendor\nesbot\carbon\src\Carbon\Traits\Creator.php:201
[Tue May  5 17:20:21 2020] PHP  19. Carbon\Carbon->change() vendor\nesbot\carbon\src\Carbon\Traits\Creator.php:177
[Tue May  5 17:20:21 2020] PHP  20. Carbon\Carbon->modify() vendor\nesbot\carbon\src\Carbon\Traits\Modifiers.php:460
[Tue May  5 17:20:21 2020] PHP  21. Carbon\Carbon->modify() \vendor\nesbot\carbon\src\Carbon\Traits\Modifiers.php:434

The library does not catch and then trow again the Exception in \Carbon\Traits\Creator::rawParse .
The function \Carbon\Traits\Modifiers::change is giving php error.
Edit : php warning:
 Warning: DateTime::modify(): Failed to parse time string (2020-04-110) at position 10 (0): Unexpected character in ...\vendor\nesbot\carbon\src\Carbon\Traits\Modifiers.php on line 434

```

Thanks!

**Comment by kylekatarnls** [comment]
I can't reproduce, I get the error `Error: DateTime::__construct(): Failed to parse time string (2020-04-110) at position 10 (0): Unexpected character` which is the expected one.

Please check you didn't mute PHP errors.

**Comment by kylekatarnls** [comment]
I can reproduce your error with:

```php
$date = @Carbon::parse('2020-04-110');

var_dump($date->format('Y-m-d'));
```

But in this case, error is muted on purpose, so it's the expected behavior. So my hypothesis is you set error_reporting on a lower level than the recommended default for development or you have an handler that caught/muted the error.

In both cases, it means Carbon behave as expected.

**Comment by kylekatarnls** [comment]
If it's not an issue with your settings, please provide a way to reproduce on https://try-carbon.herokuapp.com/ or detailed environment config and I will re-open if there is something to fix in the library.

**Comment by mariuspanteap4b** [comment]
Hi. Thanks for the quick reply. 
PHP 7.2.20 (cli) (built: Jul  2 2019 13:00:53) ( NTS MSVC15 (Visual C++ 2017) x64 )
I can't reproduce it either in your test environment...
Local, I tried also
```
        error_reporting(E_ALL);
        try{
            $test = Carbon::parse('2020-04-110')->format('Y-m-d');
            var_dump($test);
        } catch (Exception $e) {
        
          var_dump('catched ok');
        }
```
=>
Warning: DateTime::modify(): Failed to parse time string (2020-04-110) at position 10 (0): Unexpected character in ...\vendor\nesbot\carbon\src\Carbon\Traits\Modifiers.php on line 434

{"exception":"Error","message":"Call to a member function format() on boolean","code":0}

17 0.3207 3885888 Carbon\Carbon::parse( ) ...\Term.php:92
18 0.3207 3885720 Carbon\Carbon::rawParse( ) ...\Creator.php:201
19 0.3393 3944968 Carbon\Carbon->change( ) ...\Creator.php:177
20 0.3394 3944968 Carbon\Carbon->modify( ) ...\Modifiers.php:460
21 0.3394 3944968 modify ( ) ...\Modifiers.php:434

**Comment by mariuspanteap4b** [comment]
Also I modified your library to see if it works in the right manner.

```
    public static function rawParse($time = null, $tz = null)
    {
        if ($time instanceof DateTimeInterface) {
            return static::instance($time);
        }

        try {
            return new static($time, $tz);
        } catch (Exception $exception) {
            try {var_dump(1);
                return static::now($tz)->change($time);
            } catch (Exception $ignored) { var_dump(2);
                throw $exception;
            }
        }
    }
```

=>

Only 1 gets printed. 2 is NOT printed.

**Comment by mariuspanteap4b** [comment]
FYI: PHP 7.3.10 gives the same behaviour.

**Comment by mariuspanteap4b** [comment]
I can;t attach here an archive, but if you create a project:

run.php:

```
<?php

use Carbon\Carbon;

require_once 'vendor/autoload.php';

try {
    $test = Carbon::parse('2020-04-110');
    var_dump($test); //this gets printed
    $test->format('Y-m-d');// this gets executed as false->format();
} catch (Exception $e) { //Does not enter here
    var_dump('catched ok');
}
```
composer json:
```

{
  "name": ".../carbon",
  "description": "Testing CARBON library",
  "require": {
    "php": "7.2.20",
    "nesbot/carbon": "2.33.0"
  }
}
```
Run
` composer install` 
and then
`php run.php`
 you will replicate the issue.
You can test with all php versions I mentioned above.

**Comment by kylekatarnls** [comment]
What you get here: https://github.com/briannesbitt/Carbon/issues/2075#issuecomment-624452768

Is the expected behavior, the first error you get is:
```
Warning: DateTime::modify(): Failed to parse time string (2020-04-110) at position 10 (0)
```

And if you don't stop the execution (with an error handler that stop on warning and notice), then it continue to run and can raise many errors that comes from the first one.

But `Carbon::parse('2020-04-110')->format()` is a composition of parse() that fails with the expected error and so you get false, then `->format()` since the execution didn't stop you're running `false->format()`, and as you can see this is not in Carbon hands, because you're not at all calling Carbon format method, you're trying to call a method on `false`. And the reason you get `false` is well explained by the warning triggered by `DateTime`: `Failed to parse time string (2020-04-110) at position 10 (0)`

I don't see any relevant reason to wrap this `DateTime` error into an other `Carbon` exception, it properly does its job.

If you want to catch this warning use `catch(Throwable` instead of `catch(Exception`.

But you should rather do `if($test)` before calling `->format()`, and even better: you should check the string `'2020-04-110'` before trying to pass it to `Carbon::parse` it's always better to check input before processing it than letting your whole code running and failing and try to catch all possible failures.

**Comment by mariuspanteap4b** [comment]
To be honest I never thought of using Throwable... Maybe because the parse's dockblock is not accurate.
Thanks for your time.
PS.
Maybe if you would update the dockblock, people would use this function more easily.

Ex:


```
    /**
     * Create a carbon instance from a string.
     *
     * This is an alias for the constructor that allows better fluent syntax
     * as it allows you to do Carbon::parse('Monday next week')->fn() rather
     * than (new Carbon('Monday next week'))->fn().
     *
     * @param string|null              $time
     * @param DateTimeZone|string|null $tz
     *
     * @return static | false
     * @throws \Throwable
     */
```

**Comment by cfv1000** [comment]
@kylekatarnls 

Consideting @mariuspanteap4b's code, when calling `Carbon::parse('2020-04-110')->format(DATE_ATOM)`, I don't believe method `format` should ever be called. If parsing fails, we should throw an exception and stop execution. 

The main problem here is that this line is trying to call ->format on a boolean type.
Maybe the "parse" problem is that it has multiple return data types

Can we consider this change? 

Thank you

**Comment by kylekatarnls** [comment]
> if you would update the dockblock

> Can we consider this change

Feel free to propose pull-requests. If it pass the unit tests (is not a breaking change) and add value over the current error already raised, I will merge it.

**Comment by kylekatarnls** [comment]
You can try the new behavior requiring `"nesbot/carbon": "2.x-dev"`

**Comment by opensourcerules** [comment]
I saw your commit and I tested locally. From my side it's ok. Thank you.

 Bear in mind that  in php 8 @ will not suppress fatal errors anymore. 
**I don't know about warnings.**

Quote:
"The @ operator no longer silences fatal errors

It's possible that this change might reveal errors that again were hidden before PHP 8. Make sure to set display_errors=Off on your production servers!"

https://stitcher.io/blog/new-in-php-8
